README.md
# Jekyll Cloudinary Liquid tag
[![RuboCop](https://github.com/mavaddat/jekyll-cloudinary/actions/workflows/rubocop.yml/badge.svg)](https://github.com/mavaddat/jekyll-cloudinary/actions/workflows/rubocop.yml)
<!--
[![Gem Version](https://badge.fury.io/rb/jekyll-cloudinary.svg)](https://badge.fury.io/rb/jekyll-cloudinary)
[![Gem Downloads](https://img.shields.io/gem/dt/jekyll-cloudinary.svg?style=flat)](http://rubygems.org/gems/jekyll-cloudinary) -->

`jekyll-cloudinary` is a [Jekyll](http://jekyllrb.com/) plugin adding a [Liquid](http://liquidmarkup.org) tag to ease the use of [Cloudinary](https://nho.io/cloudinary-signup) for responsive images in your Markdown/[Kramdown](http://kramdown.gettalong.org/) posts.

It builds the HTML for responsive images in the posts, using the `srcset` and `sizes` attributes for the `<img />` tag (see [the "varying size and density" section of this post](https://jakearchibald.com/2015/anatomy-of-responsive-images/#varying-size-and-density) if this is new for you, and why it's recommended to [not use `<picture>` most of the time](https://cloudfour.com/thinks/dont-use-picture-most-of-the-time/)). URLs in the `srcset` are cloudinary URLs that [fetch on-the-fly](http://cloudinary.com/features#fetch) the post's images and resize them to several sizes.

You are in full control of the number of generated images and their sizes, and the `sizes` attribute that helps the browser decide which image to download. See the complete configuration options for details.

Here is the general syntax of this Liquid tag:

<!-- {% raw %} -->
```liquid
{% cloudinary cloudflare.png alt="Un schéma montrant l'apport de Cloudflare" caption="Un schéma montrant l'apport de Cloudflare" loading="lazy" %}
```
<!-- {% endraw %} -->

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
## Table of contents

- [Jekyll Cloudinary Liquid tag](#jekyll-cloudinary-liquid-tag)
  - [Table of contents](#table-of-contents)
  - [Installation](#installation)
  - [Configuration](#configuration)
    - [Mandatory settings](#mandatory-settings)
    - [Optional global settings](#optional-global-settings)
      - [`only_prod` (default: `false`)](#only_prod-default-false)
      - [`verbose` (default: `false`)](#verbose-default-false)
      - [`origin_url`](#origin_url)
    - [Optional (but highly recommended) presets](#optional-but-highly-recommended-presets)
      - [Default preset](#default-preset)
      - [Additional presets](#additional-presets)
    - [Detailed preset settings](#detailed-preset-settings)
      - [`figure` (default: `auto`)](#figure-default-auto)
      - [`min_width` (default: `320`)](#min_width-default-320)
      - [`max_width` (default: `1200`)](#max_width-default-1200)
      - [`fallback_max_width` (defaut: `1200`)](#fallback_max_width-defaut-1200)
      - [`steps` (default: `5`)](#steps-default-5)
      - [`sizes` (default: `"100vw"`)](#sizes-default-100vw)
      - [`attributes` (default: none)](#attributes-default-none)
  - [Liquid tag values](#liquid-tag-values)
  - [Liquid tag attributes](#liquid-tag-attributes)
    - [Recommended attributes](#recommended-attributes)
    - [Loading attribute](#loading-attribute)
    - [Other interesting attributes](#other-interesting-attributes)
  - [Live example](#live-example)
  - [Contributing](#contributing)
  - [Do you use the plugin on a live site?](#do-you-use-the-plugin-on-a-live-site)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Installation

[Sign up **for free** on Cloudinary!](https://nho.io/cloudinary-signup) The free account should be enough for most blogs.

Add `gem 'jekyll-cloudinary'` to the `jekyll_plugin` group in your `Gemfile`:

```ruby
source 'https://rubygems.org'

gem 'jekyll'

group :jekyll_plugins do
  gem 'jekyll-cloudinary'
end
```

Then run `bundle` to install the gem.

## Configuration

### Mandatory settings

Add `cloudinary` to your `_config.yml` and your Cloudinary "Cloud name" (find it in your [Cloudinary dashboard](https://cloudinary.com/console)):

```yaml
cloudinary:
  cloud_name: <put here your Cloudinary "Cloud name">
```

### Optional global settings

You can now define some global settings

```yaml
cloudinary:
  …
  only_prod: true
  verbose: true
  origin_url: https://another-domain.com
```

#### `only_prod` (default: `false`)

When set to `true`, this setting implies that responsive image HTML and Cloudinary URLs are generated only if the environment is `production`.

For example:

- if you run `JEKYLL_ENV=production bundle exec jekyll build`, you'll get the code to deploy, with `srcset` and Cloudinary URLs.
- if you run `JEKYLL_ENV=development bundle exec jekyll serve`, you'll get code for local development, with standard `<img src="…">` code and local URLs.

[`JEKYLL_ENV=development` is the default value](https://jekyllrb.com/docs/configuration/#specifying-a-jekyll-environment-at-build-time).

If you don't set `only_prod` or set it to `false`, responsive image HTML and Cloudinary URLs are always generated, whatever the environment. jekyll-cloudinary had only this behavior before version 1.11.0.

#### `verbose` (default: `false`)

When set to `true`, this setting will show messages in the console when something goes wrong, such as:

```
[Cloudinary] Couldn't find this image to check its width: /path/to/jekyll/_site/assets/img.jpg
```

or

```
[Cloudinary] Natural width of source image 'img.jpg' (720px) in _posts/2016-06-09-post.md not enough for creating 1600px version
```

#### `origin_url`

When `origin_url` is set, **jekyll-cloudinary** will use this URL rather than `site.url` as origin of the source images.

This allows you to store your source image on a different domain than your website.

### Optional (but highly recommended) presets

You can now define the presets you need for your posts' images, starting with the default one:

#### Default preset

The default preset is the one you don't even have to mention when using the Liquid tag, and that will be used if a preset you use in the tag doesn't exist.

```yaml
cloudinary:
  …
  presets:
    default:
      min_width: 320
      max_width: 1600
      fallback_max_width: 800
      steps: 5
      sizes: "(min-width: 50rem) 50rem, 90vw"
```

This preset will generate five images 320 to 1600 pixels wide in the `srcset` and define `sizes` as `"(min-width: 50rem) 50rem, 90vw"`. The fallback image defined in the `src` will have a width of 800 pixels.

With this preset, you only have to write this in your Markdown post:

<!-- {% raw %} -->
```liquid
{% cloudinary /assets/img.jpg alt="beautiful!" %}
```
<!-- {% endraw %} -->

To get this HTML:

```html
<img
  src="http://res.cloudinary.com/<cloud_name>/image/fetch/c_limit,w_800,q_auto,f_auto/https://<your-domain>/assets/img.jpg"
  srcset="
    http://res.cloudinary.com/<cloud_name>/image/fetch/c_limit,w_320,q_auto,f_auto/https://<your-domain>/assets/img.jpg 320w,
    http://res.cloudinary.com/<cloud_name>/image/fetch/c_limit,w_640,q_auto,f_auto/https://<your-domain>/assets/img.jpg 640w
    http://res.cloudinary.com/<cloud_name>/image/fetch/c_limit,w_960,q_auto,f_auto/https://<your-domain>/assets/img.jpg 960w
    http://res.cloudinary.com/<cloud_name>/image/fetch/c_limit,w_1280,q_auto,f_auto/https://<your-domain>/assets/img.jpg 1280w
    http://res.cloudinary.com/<cloud_name>/image/fetch/c_limit,w_1600,q_auto,f_auto/https://<your-domain>/assets/img.jpg 1600w
    "
  sizes="(min-width: 50rem) 50rem, 90vw"
  alt="beautiful!"
  width="480"
  height="320"
/>
```

There is a true default `default` preset, but you're strongly encouraged to override it with your own default preset.

#### Additional presets

You can add other presets if you need several image sizes in your posts.

Here is an example for images that take only one third of the post width:

```yaml
cloudinary:
  …
  presets:
    …
    onethird:
      min_width: 110
      max_width: 535
      fallback_max_width: 300
      steps: 3
      sizes: "(min-width: 50rem) 17rem, 30vw"
      attributes:
        class: "one3rd"
        loading: "lazy"
```

To use this additional preset, you will have to write this in your Markdown post:

<!-- {% raw %} -->
```liquid
{% cloudinary onethird /assets/img.jpg %}
```
<!-- {% endraw %} -->

The generated element will also get a `class="one3rd"` that can be useful for example with this CSS:

```css
.one3rd {
  max-width: 33%;
  float: right;
  margin: 0 0 1em 1em;
}
```

### Detailed preset settings

#### `figure` (default: `auto`)

This setting lets you decide what to do when there is a `caption` attribute in the Cloudinary Liquid tag.

The value can be:

- `auto` (default): will generate a `<figure>` and `<figcaption>` only if there's a caption
- `never`: will always generate a `<img>`, losing the caption
- `always`: will always generate a `<figure>` and `<figcaption>`, even if there's no `caption` attribute

If a `<figure>` is generated and there are attributes (in the preset or the Liquid tag), they are added to the `<img>` if they are `alt`, `title` or `loading`, or to the `<figure>`.

#### `min_width` (default: `320`)

#### `max_width` (default: `1200`)

#### `fallback_max_width` (defaut: `1200`)

#### `steps` (default: `5`)

#### `sizes` (default: `"100vw"`)

#### `attributes` (default: none)

You can define attributes that will be added to all images using this preset. Attributes are added without transformation to the generated element.

You should obviously not add to preset attributes that should have different values for each image, such as `alt`, `caption`, `title`, etc.

You can set a `class`, `aria-*` attributes for enhanced accessibility, or even `data-*` attributes you would like to use later with CSS or JavaScript.

## Liquid tag values

You can use liquid variables inside the liquid tag.

For example, if you have the picture path in a `thumbnail` attribute of the YAML Front Matter, you can use it in the tag:

<!-- {% raw %} -->
```liquid
{% cloudinary {{ page.thumbnail }} alt="{{ page.title }} image" %}
```
<!-- {% endraw %} -->

## Liquid tag attributes

You can add attributes to the liquid tag, after the image path:

<!-- {% raw %} -->
```liquid
{% cloudinary onethird /assets/selfie.jpg alt="My selfie" loading="eager" %}
```
<!-- {% endraw %} -->

Just like the ones from the preset settings, inline attributes are added without transformation to the generated element.

### Recommended attributes

You should obviously define the `alt` attribute, mandatory for accessibility.

If you want the image to be inside a `figure` element, you probably also want to add a `caption` attribute. This is the only one that can act differently than other attributes, depending on [the `figure` setting](#figure-default-auto).

You can also set a `title` attribute, but there are really few use cases for it on images.

`alt`, `caption` and `title` attributes can contain Markdown.

### Loading attribute

The `loading` attribute allows you to tell the browser how you want it to load this image.

From [this article written by Addy Osmani](https://addyosmani.com/blog/lazy-loading/):

> The loading attribute allows a browser to defer loading offscreen images and iframes until users scroll near them. loading supports three values:
> - `lazy`: is a good candidate for lazy loading.
> - `eager`: is not a good candidate for lazy loading. Load right away.
> - `auto`: browser will determine whether or not to lazily load.
> 
> Not specifying the attribute at all will have the same impact as setting loading=auto.

### Other interesting attributes

You can also use attributes to add a `class`, `aria-*` attributes for enhanced accessibility, or even `data-*` attributes you would like to use later with CSS or JavaScript.

## Live example

Go to this post: [https://nicolas-hoizey.com/2016/07/tout-change-rien-ne-change.html](https://nicolas-hoizey.com/2016/07/tout-change-rien-ne-change.html).

The source Markdown is here: [https://github.com/mavaddat/nicolas-hoizey.com/blob/master/_posts/2016/07/13-tout-change-rien-ne-change/2016-07-13-tout-change-rien-ne-change.md](https://github.com/mavaddat/nicolas-hoizey.com/blob/master/_posts/2016/07/13-tout-change-rien-ne-change/2016-07-13-tout-change-rien-ne-change.md).

The content is in french, yes, but look only at the images if you don't understand.

You'll find here:

- 2 logos floating on the right of the text (or centered on smaller screens): [Jekyll](http://jekyllrb.com/) and [Cloudinary](https://nho.io/cloudinary-signup)
- 2 screenshots taking the whole width of the content: the [Cloudinary pricing table](http://cloudinary.com/pricing), and [Dareboost](https://www.dareboost.com/en/home)'s performance monitoring graph

These image types need different settings to deal with different sizes and position:

- screenshot always use the full content width, if they're wide enough
- logos are centered and take one half of the content width on small screens, and are floated and take one fourth of the content width on larger screens

This is how I use the Cloudinary Liquid tag for the Cloudinary logo and prices table screenshot:

<!-- {% raw %} -->
```liquid
{% cloudinary logo /assets/logos/cloudinary.png alt="Logo de Cloudinary" %}
{% cloudinary cloudinary-pricing.png alt="Les tarifs de Cloudinary" caption="Les tarifs de Cloudinary, dont l'offre gratuite déjà généreuse" %}
```
<!-- {% endraw %} -->

The only difference is that I explicitly use the `logo` preset for the logo. The other image uses the `default` preset.

Here is the necessary configuration for this:

```yaml
cloudinary:
  cloud_name: …
  verbose: false
  presets:
    default:
      min_width: 320
      max_width: 1600
      fallback_max_width: 800
      steps: 5
      sizes: '(min-width: 50rem) 50rem, 90vw'
      figure: always
    logo:
      min_width: 80
      max_width: 400
      fallback_max_width: 200
      steps: 3
      sizes: '(min-width: 50rem) 13rem, (min-width: 40rem) 25vw, 45vw'
      figure: never
      attributes:
        class: logo
```

It generates these HTML fragments (pretty printed here), for the logo:

```html
<img
  src="https://res.cloudinary.com/nho/image/fetch/c_limit,w_200,q_auto,f_auto/https://nicolas-hoizey.com/assets/logos/cloudinary.png"
  srcset="
    https://res.cloudinary.com/nho/image/fetch/c_limit,w_80,q_auto,f_auto/https://nicolas-hoizey.com/assets/logos/cloudinary.png 80w,
    https://res.cloudinary.com/nho/image/fetch/c_limit,w_240,q_auto,f_auto/https://nicolas-hoizey.com/assets/logos/cloudinary.png 240w,
    https://res.cloudinary.com/nho/image/fetch/c_limit,w_400,q_auto,f_auto/https://nicolas-hoizey.com/assets/logos/cloudinary.png 400w"
  sizes="
    (min-width: 50rem) 13rem,
    (min-width: 40rem) 25vw,
    45vw"
  class="logo"
  alt="Logo de Cloudinary"
  width="480"
  height="350"
/>
```

And for the screenshot:

```html
<figure>
  <img
    src="https://res.cloudinary.com/nho/image/fetch/c_limit,w_800,q_auto,f_auto/https://nicolas-hoizey.com/2016/07/cloudinary-pricing.png"
    srcset="
      https://res.cloudinary.com/nho/image/fetch/c_limit,w_320,q_auto,f_auto/https://nicolas-hoizey.com/2016/07/cloudinary-pricing.png 320w,
      https://res.cloudinary.com/nho/image/fetch/c_limit,w_640,q_auto,f_auto/https://nicolas-hoizey.com/2016/07/cloudinary-pricing.png 640w,
      https://res.cloudinary.com/nho/image/fetch/c_limit,w_960,q_auto,f_auto/https://nicolas-hoizey.com/2016/07/cloudinary-pricing.png 960w,
      https://res.cloudinary.com/nho/image/fetch/c_limit,w_1208,q_auto,f_auto/https://nicolas-hoizey.com/2016/07/cloudinary-pricing.png 1208w"
    sizes="(min-width: 50rem) 50rem, 90vw"
    alt="Les tarifs de Cloudinary"
    width="1208"
    height="561"
  />
  <figcaption>Les tarifs de Cloudinary, dont l'offre gratuite déjà généreuse</figcaption>
</figure>
```

There are only 4 version in the `srcset` here because 2 of the 5 expected sizes are larger than the source image, and are replaced by one using the native source image width.

And here are the relevant parts of the accompanying CSS (in Sass form):

```sass
article {
  figure, img {
    margin: 2em auto;
    display: block;
    max-width: 100%;
    height: auto;
  }
}

.logo {
  display: block;
  margin: 1em auto;
  max-width: 50%;
  height: auto;

  @media (min-width: 40em) {
    max-width: 25%;
    float: right;
    margin: 0 0 1em 1em;
  }
}
```

## Contributing

Thanks for your interest in contributing! There are many ways to contribute to this project. [Get started here](./CONTRIBUTING.md).

## Do you use the plugin on a live site?

Add it to [the "Sites" page of the wiki](./wiki/Sites) and please let me know on Twitter: [@mavaddat](https://twitter.com/mavaddat)


RELEASES.md
# Releases

## [v1.21](https://github.com/mavaddat/jekyll-cloudinary/releases/tag/v1.21)

Bump to Bundler 2.2.30. Update Rubocop idioms.     

## [v1.14.1](https://github.com/mavaddat/jekyll-cloudinary/releases/tag/v1.14.1)

Fix the loading attribute handling.

Add compatibility with Jekyll 4 (thanks [@DirtyF](https://github.com/DirtyF))

## [v1.14.0](https://github.com/mavaddat/jekyll-cloudinary/releases/tag/v1.14.0)

Add support for the loading attribute, for [native lazy loading](https://addyosmani.com/blog/lazy-loading/) (Thanks [@borisschapira](https://github.com/borisschapira))

Read [the documentation for the loading attribute](https://github.com/mavaddat/jekyll-cloudinary/blob/master/README.md#loading-attribute).

## [v1.13.0](https://github.com/mavaddat/jekyll-cloudinary/releases/tag/v1.13.0)

Add cross-origin support to prevent opaque responses in Service Workers.

See https://cloudfour.com/thinks/when-7-kb-equals-7-mb/

## [v1.12.4](https://github.com/mavaddat/jekyll-cloudinary/releases/tag/v1.12.4)

Improving gem summary and description to help people find it. There was no mention of "plugin" in it… 🤔

## [v1.12.3](https://github.com/mavaddat/jekyll-cloudinary/releases/tag/v1.12.3)

Strings are now immutable by default, be careful. ([ca68ba7](https://github.com/mavaddat/jekyll-cloudinary/commit/ca68ba7743b69983836b993761d1004494197795))

## [v1.12.2](https://github.com/mavaddat/jekyll-cloudinary/releases/tag/v1.12.2)

Match jekyll's coding style thanks to [@DirtyF](https://github.com/DirtyF) with a little help from [Rubocop](http://rubocop.readthedocs.io/).

## [v1.12.1](https://github.com/mavaddat/jekyll-cloudinary/releases/tag/v1.12.1)

Break early if there is no `cloud_name` in `_config.yml`.

## [v1.12.0](https://github.com/mavaddat/jekyll-cloudinary/releases/tag/v1.12.0)

Thanks to [@suprafly](https://github.com/suprafly)'s [Pull Request](https://github.com/mavaddat/jekyll-cloudinary/pull/29), you can now host your source images on an origin domain different from your website domain.

## [v1.11.0](https://github.com/mavaddat/jekyll-cloudinary/releases/tag/v1.11.0)

Thanks to [Pascal Brokmeier](https://github.com/pascalwhoop)'s [Pull Request](https://github.com/mavaddat/jekyll-cloudinary/pull/34), you can now have responsive images HTML and Cloudinary URLs generated only when you build for production.

Just make sure to:

- set the new option `only_prod` to `true`
- and set the environment to `production` before building: `JEKYLL_ENV=production bundle exec jekyll build`

## [v1.10.0](https://github.com/mavaddat/jekyll-cloudinary/releases/tag/v1.10.0)

Fixes an issue caused by Jekyll 3.8.1 introducing a change to content's path, adding an `/#excerpt` at the end in case of an excerpt.

See https://github.com/mavaddat/jekyll-cloudinary/commit/5372e37e4d31bf1934d90665692b9e14f2ac2147

## [v1.9.1](https://github.com/mavaddat/jekyll-cloudinary/releases/tag/v1.9.1)

Better warning message when the local source image is missing.

## [v1.9.0](https://github.com/mavaddat/jekyll-cloudinary/releases/tag/v1.9.0)

Get the dimensions of the picture from the source path instead of the destination one ([#19](https://github.com/mavaddat/jekyll-cloudinary/issues/19))

## [v1.8.1](https://github.com/mavaddat/jekyll-cloudinary/releases/tag/v1.8.1)

Fixes an issue with local images.

## v1.8.0

Image size detection now uses FastImage instead of RMagick to remove imagemagick dependency, thanks to [@aarongustafson](https://github.com/aarongustafson) ([#25](https://github.com/mavaddat/jekyll-cloudinary/issues/25))

## v1.7.0

It is now possible to use all effects and transformations from the Cloudinary API, thanks to [@aarongustafson](https://github.com/aarongustafson) ([#24](https://github.com/mavaddat/jekyll-cloudinary/issues/24))

## v1.4.0

Now supports sites with baseurl ([#10](https://github.com/mavaddat/jekyll-cloudinary/issues/10))

## v1.3.1

Restores natural width if some computed ones are missing.

## v1.3.0

Restores `width` and `height` attributes.

## v1.2.17

Fixes a little typo with huge impact…

## v1.2.16

Code improvements thanks to Rubocop, initialized by [@DirtyF](https://github.com/DirtyF)

## v1.2.15

Fixed bugs:
- Don’t add `<img>` `height` & `width` attributes ([#7](https://github.com/mavaddat/jekyll-cloudinary/issues/7))
- Dont’t wrap `alt` text in `<p>` ([#8](https://github.com/mavaddat/jekyll-cloudinary/issues/8))
- `require 'rmagick'` tripped me up ([#11](https://github.com/mavaddat/jekyll-cloudinary/issues/11))

Thanks [@eeeps](https://github.com/eeeps) for catching these issues!


lib/jekyll/cloudinary.rb
# frozen_string_literal: true

module Jekyll
  module Cloudinary

    class CloudinaryTag < Liquid::Tag
      require "fastimage"

      def initialize(tag_name, markup, tokens)
        @markup = markup
        super
      end

      def render(context)
        # Default settings
        settings_defaults = {
          "cloud_name"         => "",
          "only_prod"          => false,
          "verbose"            => false,
        }
        preset_defaults = {
          "min_width"          => 320,
          "max_width"          => 1200,
          "fallback_max_width" => 1200,
          "steps"              => 5,
          "sizes"              => "100vw",
          "figure"             => "auto",
          "attributes"         => {},
          "width_height"       => true,
          # Cloudinary transformations
          "height"             => false,
          "crop"               => "limit",
          "aspect_ratio"       => false,
          "gravity"            => false,
          "zoom"               => false,
          "x"                  => false,
          "y"                  => false,
          "format"             => false,
          "fetch_format"       => "auto",
          "quality"            => "auto",
          "radius"             => false,
          "angle"              => false,
          "effect"             => false,
          "opacity"            => false,
          "border"             => false,
          "background"         => false,
          "overlay"            => false,
          "underlay"           => false,
          "default_image"      => false,
          "delay"              => false,
          "color"              => false,
          "color_space"        => false,
          "dpr"                => false,
          "page"               => false,
          "density"            => false,
          "flags"              => false,
          "transformation"     => false,
        }

        # TODO: Add validation for these parameters
        transformation_options = {
          "height"         => "h",
          "crop"           => "c", # can include add-on: imagga_scale
          "aspect_ratio"   => "ar",
          "gravity"        => "g",
          "zoom"           => "z",
          "x"              => "x",
          "y"              => "y",
          "fetch_format"   => "f",
          "quality"        => "q", # can include add-on: jpegmini
          "radius"         => "r",
          "angle"          => "a",
          "effect"         => "e", # can include add-on: viesus_correct
          "opacity"        => "o",
          "border"         => "bo",
          "background"     => "b",
          "overlay"        => "l",
          "underlay"       => "u",
          "default_image"  => "d",
          "delay"          => "dl",
          "color"          => "co",
          "color_space"    => "cs",
          "dpr"            => "dpr",
          "page"           => "pg",
          "density"        => "dn",
          "flags"          => "fl",
          "transformation" => "t",
        }

        # Settings
        site = context.registers[:site]
        site_url = site.config["url"] || ""
        site_baseurl = site.config["baseurl"] || ""
        if site.config["cloudinary"].nil?
          Jekyll.logger.abort_with("[Cloudinary]", "You must set your cloud_name in _config.yml")
        end
        settings = settings_defaults.merge(site.config["cloudinary"])
        if settings["cloud_name"] == ""
          Jekyll.logger.abort_with("[Cloudinary]", "You must set your cloud_name in _config.yml")
        end
        url = settings["origin_url"] || (site_url + site_baseurl)

        # Get Markdown converter
        markdown_converter = site.find_converter_instance(::Jekyll::Converters::Markdown)

        preset_user_defaults = {}
        if settings["presets"]
          if settings["presets"]["default"]
            preset_user_defaults = settings["presets"]["default"]
          end
        end

        preset = preset_defaults.merge(preset_user_defaults)

        # Render any liquid variables in tag arguments and unescape template code
        rendered_markup = Liquid::Template
          .parse(@markup)
          .render(context)
          .gsub(%r!\\\{\\\{|\\\{\\%!, '\{\{' => "{{", '\{\%' => "{%")

        # Extract tag segments
        markup =
          %r!^(?:(?<preset>[^\s.:\/]+)\s+)?(?<image_src>[^\s]+\.[a-zA-Z0-9]{3,4})\s*(?<html_attr>[\s\S]+)?$!
            .match(rendered_markup)

        unless markup
          Jekyll.logger.abort_with("[Cloudinary]", "Can't read this tag: #{@markup}")
        end

        image_src = markup[:image_src]

        # Dynamic image type
        type = "fetch"
        # TODO: URL2PNG requires signed URLs... need to investigate more
        # if /^url2png\:/.match(image_src)
        #   type = "url2png"
        #   image_src.gsub! "url2png:", ""
        # end

        if markup[:preset]
          if settings["presets"][markup[:preset]]
            preset = preset.merge(settings["presets"][markup[:preset]])
          elsif settings["verbose"]
            Jekyll.logger.warn(
              "[Cloudinary]",
              "'#{markup[:preset]}' preset for the Cloudinary plugin doesn't exist, \
              using the default one"
            )
          end
        end

        attributes = preset["attributes"]

        # Deep copy preset for single instance manipulation
        instance = Marshal.load(Marshal.dump(preset))

        # Process attributes
        html_attr = if markup[:html_attr]
                      Hash[ *markup[:html_attr].scan(%r!(?<attr>[^\s="]+)(?:="(?<value>[^"]+)")?\s?!).flatten ]
                    else
                      {}
                    end

        if instance["attr"]
          html_attr = instance.delete("attr").merge(html_attr)
        end

        # Classes from the tag should complete, not replace, the ones from the preset
        if html_attr["class"] && attributes["class"]
          html_attr["class"] << " #{attributes["class"]}"
        end
        html_attr = attributes.merge(html_attr)

        # Deal with the "caption" attribute as a true <figcaption>
        if html_attr["caption"]
          caption = markdown_converter.convert(html_attr["caption"])
          html_attr.delete("caption")
        end

        # alt and title attributes should go only to the <img> even when there is a caption
        img_attr = "".dup
        if html_attr["alt"]
          img_attr << " alt=\"#{html_attr["alt"]}\""
          html_attr.delete("alt")
        end
        if html_attr["title"]
          img_attr << " title=\"#{html_attr["title"]}\""
          html_attr.delete("title")
        end
        if html_attr["loading"]
          img_attr << " loading=\"#{html_attr["loading"]}\""
          html_attr.delete("loading")
        end

        attr_string = html_attr.map { |a, v| "#{a}=\"#{v}\"" }.join(" ")

        # Figure out the Cloudinary transformations
        transformations = []
        transformations_string = ""
        transformation_options.each do |key, shortcode|
          if preset[key]
            transformations << "#{shortcode}_#{preset[key]}"
          end
        end
        unless transformations.empty?
          transformations_string = transformations.compact.reject(&:empty?).join(",") + ","
        end

        # Build source image URL
        is_image_remote = %r!^https?!.match(image_src)
        if is_image_remote
          # It's remote
          image_dest_path = image_src
          image_dest_url = image_src
          natural_width, natural_height = FastImage.size(image_dest_url)
          if natural_width.nil?
            Jekyll.logger.warn("remote url doesn't exists " + image_dest_url)
            return "<img src=\"#{image_dest_url}\" />"
          end
          width_height = "width=\"#{natural_width}\" height=\"#{natural_height}\""
          fallback_url = "https://res.cloudinary.com/#{settings["cloud_name"]}/image/#{type}/#{transformations_string}w_#{preset["fallback_max_width"]}/#{image_dest_url}"
        else
          # It's a local image
          is_image_src_absolute = %r!^/.*$!.match(image_src)
          if is_image_src_absolute
            image_src_path = File.join(
              site.config["source"],
              image_src
            )
            image_dest_path = File.join(
              site.config["destination"],
              image_src
            )
            image_dest_url = File.join(
              url,
              image_src
            )
          else
            image_src_path = File.join(
              site.config["source"],
              File.dirname(context["page"]["path"].chomp("/#excerpt")),
              image_src
            )
            image_dest_path = File.join(
              site.config["destination"],
              File.dirname(context["page"]["url"]),
              image_src
            )
            image_dest_url = File.join(
              url,
              File.dirname(context["page"]["url"]),
              image_src
            )
          end
          if File.exist?(image_src_path)
            natural_width, natural_height = FastImage.size(image_src_path)
            width_height = "width=\"#{natural_width}\" height=\"#{natural_height}\""
            fallback_url = "https://res.cloudinary.com/#{settings["cloud_name"]}/image/#{type}/#{transformations_string}w_#{preset["fallback_max_width"]}/#{image_dest_url}"
          else
            natural_width = 100_000
            width_height = ""
            Jekyll.logger.warn(
              "[Cloudinary]",
              "Couldn't find this image to check its width: #{image_src_path}."
            )
            fallback_url = image_dest_url
          end
        end

        # Don't generate responsive image HTML and Cloudinary URLs for local development
        if settings["only_prod"] && ENV["JEKYLL_ENV"] != "production"
          return "<img src=\"#{image_dest_url}\" #{attr_string} #{img_attr} #{width_height} crossorigin=\"anonymous\" />"
        end

        srcset = []
        steps = preset["steps"].to_i
        min_width = preset["min_width"].to_i
        max_width = preset["max_width"].to_i
        step_width = (max_width - min_width) / (steps - 1)
        sizes = preset["sizes"]

        if natural_width < min_width
          if settings["verbose"]
            Jekyll.logger.warn(
              "[Cloudinary]",
              "Width of source image '#{File.basename(image_src)}' (#{natural_width}px) \
              in #{context["page"]["path"]} not enough for ANY srcset version"
            )
          end
          srcset << "https://res.cloudinary.com/#{settings["cloud_name"]}/image/#{type}/#{transformations_string}w_#{natural_width}/#{image_dest_url} #{natural_width}w"
        else
          missed_sizes = []
          (1..steps).each do |factor|
            width = min_width + (factor - 1) * step_width
            if width <= natural_width
              srcset << "https://res.cloudinary.com/#{settings["cloud_name"]}/image/#{type}/#{transformations_string}w_#{width}/#{image_dest_url} #{width}w"
            else
              missed_sizes.push(width)
            end
          end
          unless missed_sizes.empty?
            srcset << "https://res.cloudinary.com/#{settings["cloud_name"]}/image/#{type}/#{transformations_string}w_#{natural_width}/#{image_dest_url} #{natural_width}w"
            if settings["verbose"]
              Jekyll.logger.warn(
                "[Cloudinary]",
                "Width of source image '#{File.basename(image_src)}' (#{natural_width}px) \
                in #{context["page"]["path"]} not enough for #{missed_sizes.join("px, ")}px \
                version#{missed_sizes.length > 1 ? "s" : ""}"
              )
            end
          end
        end
        srcset_string = srcset.join(",\n")

        # preset['figure'] can be 'never', 'auto' or 'always'
        if (caption || preset["figure"] == "always") && preset["figure"] != "never"
          "\n<figure #{attr_string}>\n<img src=\"#{fallback_url}\" srcset=\"#{srcset_string}\" sizes=\"#{sizes}\" #{img_attr} #{width_height} />\n<figcaption>#{caption}</figcaption>\n</figure>\n"
        else
          "<img src=\"#{fallback_url}\" srcset=\"#{srcset_string}\" sizes=\"#{sizes}\" #{attr_string} #{img_attr} #{width_height} crossorigin=\"anonymous\" />"
        end
      end
    end

  end
end

Liquid::Template.register_tag("cloudinary", Jekyll::Cloudinary::CloudinaryTag)
